File: server/ruoyi-modules/jshpos-order/src/main/java/com/jingshanghui/pos/order/domain/PromotedOrderSnapshotCodec.java
package com.jingshanghui.pos.order.domain;

import com.jingshanghui.pos.foundation.domain.CanonicalJson;
import com.jingshanghui.pos.order.application.model.PromotedOrderCommands.PromotedLine;
import com.jingshanghui.pos.order.application.model.PromotedOrderCommands.SubmitPromotedCashOrder;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** ORD-003 与 Flutter 共用字段、排序和摘要语义的订单快照编码器。 */
public final class PromotedOrderSnapshotCodec {

    public static final int MAX_BYTES = 1024 * 1024;

    private PromotedOrderSnapshotCodec() {
    }

    /**
     * 只编码已冻结事实，不执行任何促销算法。
     *
     * @param command 已验证的含促销订单命令
     * @param trustedCashierId 服务端可信收银员ID
     * @return 规范化JSON与摘要
     */
    public static CanonicalJson.Result encode(SubmitPromotedCashOrder command, Long trustedCashierId) {
        List<Map<String, Object>> lines = new ArrayList<>();
        command.lines().stream().sorted(Comparator.comparingInt(PromotedLine::lineNo)).forEach(line -> {
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("lineId", line.lineId()); item.put("lineNo", line.lineNo());
            item.put("skuId", line.skuId().toString()); item.put("skuCode", line.skuCode());
            if (line.barcode() != null) item.put("barcode", line.barcode());
            item.put("productName", line.productName()); item.put("unitId", line.unitId().toString());
            item.put("unitCode", line.unitCode());
            item.put("quantity", OrderRules.requireQuantity(line.quantity()).toPlainString());
            item.put("unitPriceMinor", line.unitPriceMinor()); item.put("grossAmountMinor", line.grossAmountMinor());
            item.put("discountAmountMinor", line.discountAmountMinor());
            item.put("surchargeAmountMinor", line.surchargeAmountMinor());
            item.put("payableAmountMinor", line.payableAmountMinor()); item.put("priceSource", line.priceSource());
            item.put("sourceAllocations", line.sourceAllocations());
            if (line.measuredBarcodeSnapshot() != null) {
                item.put("measuredBarcodeSnapshot", line.measuredBarcodeSnapshot().toCanonicalMap());
            }
            lines.add(item);
        });
        Map<String, Object> value = new LinkedHashMap<>();
        value.put("schemaVersion", 2); value.put("orderId", command.orderId());
        value.put("storeId", command.storeId().toString()); value.put("terminalId", command.terminalId());
        value.put("shiftId", command.shiftId()); value.put("cashierId", trustedCashierId.toString());
        value.put("businessDate", command.businessDate().toString()); value.put("storeTimezone", command.storeTimezone());
        value.put("currency", "CNY"); value.put("grossAmountMinor", command.grossAmountMinor());
        value.put("discountAmountMinor", command.discountAmountMinor());
        value.put("surchargeAmountMinor", command.surchargeAmountMinor());
        value.put("receivableAmountMinor", command.receivableAmountMinor());
        value.put("catalogVersion", command.catalogVersion()); value.put("priceVersion", command.priceVersion());
        value.put("industryTemplateVersion", command.industryTemplateVersion());
        value.put("promotionSnapshotId", command.promotionSnapshotId());
        value.put("promotionSnapshotHash", "sha256:" + command.promotionSnapshotSha256());
        value.put("quoteFingerprint", command.quoteFingerprint());
        value.put("settlementFingerprint", command.settlementFingerprint());
        value.put("promotionPackageVersion", command.promotionPackageVersion());
        value.put("manualEventRefs", command.manualEventRefs()); value.put("lines", lines);
        return CanonicalJson.from(value, MAX_BYTES);
    }
}

File: pos-flutter/lib/features/checkout/domain/promoted_order_snapshot_codec.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import 'checkout_models.dart';

/// POS-006 与 ORD-003 共用的订单快照规范化编码器。
abstract final class PromotedOrderSnapshotCodec {
  /// 构造只包含冻结事实的订单快照，不执行促销计算。
  static Map<String, Object?> document({
    required PromotedCashSaleCommand command,
    required TrustedDeviceBinding binding,
    required String promotionSnapshotSha256,
  }) => {
    'schemaVersion': 2,
    'orderId': command.basket.orderId,
    'storeId': binding.storeId,
    'terminalId': binding.terminalId,
    'shiftId': command.shiftId,
    'cashierId': binding.cashierId,
    'businessDate': command.businessDate,
    'storeTimezone': binding.storeTimezone,
    'currency': 'CNY',
    'grossAmountMinor': command.grossAmountMinor,
    'discountAmountMinor': command.discountAmountMinor,
    'surchargeAmountMinor': command.surchargeAmountMinor,
    'receivableAmountMinor': command.receivableAmountMinor,
    'catalogVersion': command.catalogVersion,
    'priceVersion': command.priceVersion,
    'industryTemplateVersion': command.industryTemplateVersion,
    'promotionSnapshotId': command.promotionSnapshotId,
    'promotionSnapshotHash': 'sha256:$promotionSnapshotSha256',
    'quoteFingerprint': command.quoteFingerprint,
    'settlementFingerprint': command.settlementFingerprint,
    'promotionPackageVersion': command.packageVersion,
    'manualEventRefs': command.manualEventRefs,
    if (command.memberBenefitSnapshot != null)
      'memberBenefitSnapshot': command.memberBenefitSnapshot!.toJson(),
    'lines': command.lines.map((line) => line.toSnapshot()).toList(),
  };

  /// 对对象键递归排序；数组保持业务顺序，与服务端 CanonicalJson 一致。
  static String canonicalJson(Object? value) => jsonEncode(_sortJson(value));

  /// 生成不带前缀的小写 SHA-256 十六进制摘要。
  static String sha256Hex(Object? value) =>
      sha256.convert(utf8.encode(canonicalJson(value))).toString();

  static Object? _sortJson(Object? value) {
    if (value is Map) {
      final normalized = <String, Object?>{
        for (final entry in value.entries) '${entry.key}': entry.value,
      };
      final keys = normalized.keys.toList()..sort();
      return <String, Object?>{
        for (final key in keys) key: _sortJson(normalized[key]),
      };
    }
    if (value is Iterable) return value.map(_sortJson).toList();
    return value;
  }
}
